File: Code/Python/Walker.py
import numpy as np
import numpy.random as rng

class Walker:
	"""
	Class defining a walker, and the scalar functions
	(log-likelihood or -energy) that you want to increase
	"""
	def __init__(self):
		"""
		Constructor: does nothing.
		"""
		pass

	def from_prior(self):
		"""
		Generate coordinates/parameters from the prior
		"""
		self.x = rng.rand(100)

	def proposal(self):
		"""
		Do something to generate a Metropolis proposal
		"""
		# Choose a probability from log p ~ uniform
		p = 10.**(-3.*rng.rand())

		# Move each coordinate with probability p
		which = rng.rand(len(self.x)) < p

		# Make sure we're at least moving one
		num_moving = which.sum()
		if num_moving == 0:
			which[rng.randint(len(self.x))] = True
			num_moving = 1

		# Move using a heavy-tailed proposal
		self.x[which] += 10.**(1.5 - 6.*rng.rand(num_moving))\
								*rng.randn(num_moving)

		# Mod into prior range
		self.x = np.mod(self.x, 1.)

		# Return zero -- in general return log of
		# a factor that enforces detailed balance
		# wrt the prior.
		return 0.

	@property
	def scalars(self):
		"""
		Calculate the scalar functions of interest
		and return them in the form of a numpy array.
		Current demo problem: the first scalar wants
		the coordinates to cluster in a gaussian in
		the center of the domain, and the second wants
		them to have periodic density. The coordinates
		are IID in this problem so it's easy to numerically
		compute the true partition function.
		"""
		s = np.empty(2)
		s[0] = -0.5*np.sum(self.x**2)
		s[1] = -np.sum(np.sin(4.*np.pi*self.x)**2)
		return s

File: Code/Julia/Walker.jl
include("Utils.jl")

# Walker class. Defines what problem we're solving.
# An object of this class is a point in parameter space. 
type Walker
	# Number of dimensions
	N::Int64

	# The parameters
	params::Array{Float64, 1}

	# The scalars
	scalars::Array{Float64, 1}
end

# Constructor that takes no input
function Walker()
	# Set the number of dimensions, the parameters, and the scalars
	Walker(100, zeros(100), zeros(2))
end

# Initialise a walker from the prior
function from_prior!(walker::Walker)
	walker.params = rand(walker.N)
	calculate_scalars!(walker)
end

# Calculate the scalars
function calculate_scalars!(walker::Walker)
	walker.scalars[1] = -sum((walker.params - 0.5).^2)
	walker.scalars[2] = -sum(4.*pi*walker.params.^2)
end

# Move a Walker according to the prior
function proposal!(walker::Walker)
	# Do something to generate a Metropolis proposal
	# Choose a probability from log p ~ uniform
	p = 10.0^(-3.0*rand())

	# Move each coordinate with probability p
	which = rand(walker.N) .< p

	# Make sure we're at least moving one
	num_moving = sum(which)
	if num_moving == 0
		which[rand(1:walker.N)] = true
		num_moving = 1
	end

	# Move using a heavy-tailed proposal
	walker.params[which] += randh(num_moving)

	# Mod into prior range
	walker.params = mod(walker.params, 1.0)

	# Calculate scalars
	calculate_scalars!(walker)

	# Return zero.
	# In general return log of a factor that enforces detailed balance
	# wrt the prior.
	return 0.0
end


